sources/softwareupdate.cpp

```diff
@@ -0,0 +1,239 @@
+/******************************************************************************
+ *
+ * Copyright (C) 2018-2020 Marton Borzak <hello@martonborzak.com>
+ *
+ * This file is part of the YIO-Remote software project.
+ *
+ * YIO-Remote software is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * YIO-Remote software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with YIO-Remote software. If not, see <https://www.gnu.org/licenses/>.
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *****************************************************************************/
+
+#include "softwareupdate.h"
+
+static Q_LOGGING_CATEGORY(CLASS_LC, "softwareupdate");
+
+SoftwareUpdate *SoftwareUpdate::s_instance = nullptr;
+
+SoftwareUpdate::SoftwareUpdate(bool autoUpdate, QString appPath, BatteryFuelGauge *batteryFuelGauge, QObject *parent)
+    : QObject(parent), m_batteryFuelGauge(batteryFuelGauge), m_autoUpdate(autoUpdate), m_appPath(appPath) {
+    s_instance = this;
+
+    m_downloadTimer = new QElapsedTimer();
+
+    // start update checker timer
+    m_checkForUpdateTimer = new QTimer(this);
+    m_checkForUpdateTimer->setInterval(3600000);  // 1 hour
+    connect(m_checkForUpdateTimer, &QTimer::timeout, this, &SoftwareUpdate::onCheckForUpdateTimerTimeout);
+    m_checkForUpdateTimer->start();
+
+    qCDebug(CLASS_LC) << "Current version:" << m_currentVersion;
+
+    // check for update on startup as well
+    checkForUpdate();
+}
+
+SoftwareUpdate::~SoftwareUpdate() {
+    s_instance = nullptr;
+    m_checkForUpdateTimer->stop();
+    m_checkForUpdateTimer->deleteLater();
+}
+
+QObject *SoftwareUpdate::getQMLInstance(QQmlEngine *engine, QJSEngine *scriptEngine) {
+    Q_UNUSED(scriptEngine)
+    Q_ASSERT(s_instance);
+
+    QObject *instance = s_instance;
+    engine->setObjectOwnership(instance, QQmlEngine::CppOwnership);
+    return instance;
+}
+
+void SoftwareUpdate::onCheckForUpdateTimerTimeout() { checkForUpdate(); }
+
+void SoftwareUpdate::checkForUpdate() {
+    m_manager = new QNetworkAccessManager(this);
+    QNetworkRequest request;
+
+    connect(m_manager, &QNetworkAccessManager::finished, this, &SoftwareUpdate::checkForUpdateFinished);
+
+    request.setUrl(QUrl(m_updateUrl));
+    request.setRawHeader("Content-Type", "application/json");
+    request.setRawHeader("Accept", "application/json");
+
+    QJsonObject json;
+    json.insert("version", m_currentVersion);
```

Comment:
We should also add a device type parameter to the update request.
Otherwise we cannot distinguish a real YIO remote from all the test and desktop installations.
E.g. `?type=remote` if running on the YIO remote or `?type=macos|win|linux` elsewhere. Or something similar...